### DISCUSSION TITLE
Animate plots

### DISCUSSION THREAD

**Discussion by IsraelGebresilasieKimo** [jddingemanse/awtiCodeDev#18] [Current code]
I have uploaded as sample a ipynb file (Precipitation_Forecast_GFS.ipynb). There is already some tempAnimation script, but how to use it for my code is unclear. Is it possible to turn the created plots into an animation; can anyone help me on the code?

**Comment by jddingemanse** [comment]
Hi Israel,

I have checked your code; based on your example I have managed to make an animation, but with a slightly different type of visualisation (pcolormesh instead of contourf; I do not yet know how to animate a contourf).

Besides that, I have some suggestions for loading the code, and I have implemented that only data for Ethiopia is shown. I do not use Basemap, instead I use GeoPandas for selecting the points within the polygon of Ethiopia, and for the rest I only use xarray (for working with the netcdf data) and matplotlib for the plotting.

### Data preparation
Import the necessary packages
```
import matplotlib.pyplot as plt
from matplotlib import animation
import numpy as np
from datetime import datetime,timedelta
from siphon.catalog import TDSCatalog
import xarray as xr
import geopandas as gpd
```
Downloading the data for 3 days. You did them one by one, I now use a loop. Also, I have added that the data is saved locally as a netcdf file (for some reason, your approach of directly using the downloaded data as bytes in the xr.open_dataset() function did not work for me).
```
Dta_URL = ('http://thredds.ucar.edu/thredds/catalog/grib/NCEP/GFS/'
              'Global_0p25deg/catalog.xml?dataset=grib/NCEP/GFS/Global_0p25deg/Best')

best_gfs = TDSCatalog(Dta_URL)
print(list(best_gfs.datasets))

best_ds = best_gfs.datasets[0]


ncss = best_ds.subset()
query = ncss.query()

filenames = ['GFSday1.nc','GFSday2.nc','GFSday3.nc']

for i,filename in enumerate(filenames):
    query.lonlat_box(north=15, south=3, east=49, west=32.5).time_range(datetime.utcnow()+ timedelta(days=i), 
                                                                   datetime.utcnow() + timedelta(days=i+1))
    query.accept('netcdf4')
    query.variables('Precipitation_rate_surface')
    data = ncss.get_data_raw(query)

    local_filename = 'output/'+filename
    print(local_filename)
    file = open(local_filename, 'wb')
    file.write(data)
    file.close()
```
Load the locally saved netcdf files, and with geopandas open shapefiles of the world and of Ethiopia (I have uploaded these shapefiles to the folder /data).
```
gfs1 = xr.open_dataset('output/'+filenames[0])
gfs2 = xr.open_dataset('output/'+filenames[1])
gfs3 = xr.open_dataset('output/'+filenames[2])

ET = gpd.read_file('data/ETadm0.zip')
world=gpd.read_file('data/world-administrative-boundaries.zip')
```
Then, for each of the three datasets, I iteratively
- Follow your steps for getting daily values (so, only remaining with lat/lon as dimensions)
- Turn the xarray Dataset into a DataFrame, with the latitude and longitude not as index but as columns (.reset_index)
- Turn the latitudes and longitudes into geographical coordinates (gpd.points_from_xy)
- Set data for all points not within the shape of ET to nan
- Turn the resulting DataFrame back into an xarray Dataset, again with lat/lon as dimensions (if you turn a Dataframe into a Dataset with `.to_xarray()`, it will turn the columns into data_vars and the index into dims. So, you can turn your latitude and longitude columns into index, and then turn that DataFrame into a Dataset).

```
gfslist = [gfs1,gfs2,gfs3]
RFsumlist = []

for gfs in gfslist:
    RF = gfs.Precipitation_rate_surface*86400
    RFsum = RF.sum(dim='time1')
    RFdf = RFsum.to_dataframe().reset_index()
    coords = gpd.points_from_xy(RFdf.longitude,RFdf.latitude,crs='EPSG:4326')
    RFdf.loc[~(coords.within(ET.geometry[0])),'Precipitation_rate_surface'] = np.nan
    RFds = RFdf.set_index(['latitude','longitude']).to_xarray()
    RFsumlist.append(RFds)    

RF1,RF2,RF3 = RFsumlist
```

### Plotting one of them with contourf
You have used Basemap. I am not familiar with Basemap. I do however think that you can approach the same result by simply using matplotlib plotting. Below, the code for a filled contour plot for the first Dataset (RF1). I also plot the world boundaries; however, as soon as you plot those the limits of your x- and y-axis will go to the full world, so before that you have to save your xlim and ylim and then reset them.
```
fig,ax=plt.subplots(figsize=(7,7))
contour = ax.contourf(RF1.longitude,RF1.latitude,RF1.Precipitation_rate_surface)
ax.grid(ls=':',lw=0.5)
xlim = ax.get_xlim()
ylim = ax.get_ylim()
world.plot(ax=ax,facecolor='none')
ax.set_xlim(xlim)
ax.set_ylim(ylim)
fig.colorbar(contour,location='right',pad=0.03)
```
For animation, I could not yet find how to do it for this contourf. The key to animation (as far as I know it, and that is not a lot) is to make one plot, and then to change the actual values in that plot. I do not know how to get the values of the contourf plot. I do however know how it works for pcolormesh. pcolormesh nicely saves all the color values in an array; you can change those with .set_array(). So, below the code for animating the three datasets with a pcolormesh.
```
titlelist = ['Day 1','Day 2','Day 3']
fig, ax = plt.subplots()
cax = ax.pcolormesh(RFsumlist[0].longitude, RFsumlist[0].latitude, RFsumlist[0].Precipitation_rate_surface, cmap='coolwarm',vmax=180)
ax.grid(ls=':',lw=0.5)
xlim = ax.get_xlim()
ylim = ax.get_ylim()
world.plot(ax=ax,facecolor='none')
ax.set_xlim(xlim)
ax.set_ylim(ylim)
fig.colorbar(cax,location='right',pad=0.03)
ax.set_title(titlelist[0])

def animate(i):
   cax.set_array(np.array(RFsumlist[i].Precipitation_rate_surface).flatten())
   ax.set_title(titlelist[i])

anim = animation.FuncAnimation(fig, animate, interval=1000, frames=3)
```
For animation, you start with making a plot. Then, in a function you write what should change each frame. In this case. the values of the pcolormesh (saved as `cax`, changing them with `set_array`) should change to the data of the next day. The different datasets are stored in `RFsumlist`, so each time we get the next one from that list. Additionally, I have added a title to distinguish between the separate days.
Setting frames to 3 makes sure that i will go over 0,1,2 --> the number of frames is of course equal to the number of datasets in our list.

I hope this helps!

**edit 30/12/2022**: I have now also found how to animate contourf. You can try it with the following code: (I am not fully happy with the colorbar yet, but at least the animation should work)
```
titlelist = ['Day 1','Day 2','Day 3']
fig, ax = plt.subplots()
cax = ax.contourf(RFsumlist[0].longitude, RFsumlist[0].latitude, RFsumlist[0].Precipitation_rate_surface, cmap='coolwarm',vmax=180)
ax.grid(ls=':',lw=0.5)
xlim = ax.get_xlim()
ylim = ax.get_ylim()
world.plot(ax=ax,facecolor='none')
ax.set_xlim(xlim)
ax.set_ylim(ylim)
fig.colorbar(cax,location='right',pad=0.03)
ax.set_title(titlelist[0])

def animate(i):
    global cax
    for c in cax.collections:
        c.remove()
    cax = ax.contourf(RFsumlist[i].longitude, RFsumlist[i].latitude, RFsumlist[i].Precipitation_rate_surface)
    ax.set_title(titlelist[i])

anim = animation.FuncAnimation(fig, animate, interval=1000, frames=3)
fig.show()
```

**Reply by IsraelGebresilasieKimo** [reply]
Dear Jan,
Thank you so much I did everything step by step it is a great amendment and I will work for further enhancement. 
Thank you again.

**Comment by IsraelGebresilasieKimo** [comment]
Dear Jan,

Thank you so much. I did everything step by step. It is a great amendment
and I will work for further enhancement.

Thank you again.

*With Regards, *

*************************************************

*Israel Gebresilasie (MSc in Climate Change and Development)*


*Lecturer, Faculty of Meteorology and Hydrology, Water Technology
Institute, Arba Minch University*



*Arba Minch Ethiopia ******************************* Phone: +251 927833912*

*Alternate email: ***@***.***
***@***.***>*

**Reply by jddingemanse** [reply]
I have edited by post by adding also code for animating contourf. Let me know whether it works!

**Comment by jddingemanse** [comment]
Speaking about animations; I have also updated the function for animating temperatures on lat/lon map under the package awtiCode. I am curious whether it works for you. You should reinstall the package awtiCode to get the newest version:
- In the Anaconda Prompt:
`pip install git+https://www.github.com/jddingemanse/awtiCode.git`

After succesfull installation, you can try the function with some .nc file that has temperatures with dimensions time, latitude and longitude. For example, from the training, the file m5_temperatures.nc, try out the following code:
```
import xarray as xr
from awtiCode import tempAnimation as ta

ds = xr.open_dataset('data/m5_temperatures.nc')
animation = ta.animateLatLonNC(ds,data_var='air')
```
Or, only animating around Ethiopia:
```
import xarray as xr
import pandas as pd
from awtiCode import tempAnimation as ta

ds = xr.open_dataset('data/m5_temperatures.nc')

latmin,latmax,lonmin,lonmax=3,16,32,48
dsET = ds.air[:,(ds.lat>latmin)&(ds.lat<latmax),(ds.lon>lonmin)&(ds.lon<lonmax)]
animation = ta.animateLatLonNC(dsET)
```
